pkg/audit/reporter.go

```diff
@@ -0,0 +1,96 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package audit
+
+import (
+	"context"
+
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/provider"
+	dbx "storj.io/storj/pkg/statdb/dbx"
+	proto "storj.io/storj/pkg/statdb/proto"
+	"storj.io/storj/pkg/statdb/sdbclient"
+)
+
+type reporter interface {
+	RecordFailedAudits(ctx context.Context, failedNodes []*pb.Node) (err error)
+}
+
+// Reporter records audit reports in statdb and implements the reporter interface
+type Reporter struct {
+	statdb sdbclient.Client
+}
+
+var (
+	port   = ":7777"
+	apiKey = []byte("")
+)
+
+// NewReporter instantiates a reporter
+func NewReporter(ctx context.Context) (reporter *Reporter, err error) {
+	ca, err := provider.NewCA(ctx, 12, 14)
+	if err != nil {
+		return nil, err
+	}
+	identity, err := ca.NewIdentity()
+	if err != nil {
+		return nil, err
+	}
+	client, err := sdbclient.NewClient(identity, port, apiKey)
+	if err != nil {
+		return nil, err
+	}
+	return &Reporter{statdb: client}, nil
+}
+
+// RecordFailedAudits saves failed audit details to statdb
+func (reporter *Reporter) RecordFailedAudits(ctx context.Context, failedNodes []*pb.Node) (err error) {
+	nodes, err := reporter.recordEntryIfNotExists(ctx, failedNodes)
+	if err != nil {
+		return err
+	}
+
+	finished := false
+	retries := 0
+	for !finished && retries < 3 {
+		_, failedNodes, err := reporter.statdb.UpdateBatch(ctx, nodes)
+		if err != nil {
+			return err
+		}
+		if len(failedNodes) == 0 {
+			finished = true
+		}
+		nodes = failedNodes
+		retries++
+	}
+	if retries == 3 {
+		return Error.New("some nodes who failed the audit also failed to be updated in statdb")
+	}
+	return nil
+}
+
+// creates a statdb proto with node information and saves to statdb if it didn't already exist
+func (reporter *Reporter) recordEntryIfNotExists(ctx context.Context, failedNodes []*pb.Node) (nodes []*proto.Node, err error) {
+	nodes = make([]*proto.Node, len(failedNodes))
+	for i, fail := range failedNodes {
+		nodes[i] = &proto.Node{
+			NodeId:             []byte(fail.GetId()),
```

Comment:
`NodeId` is the only thing you need here if you are just doing `createEntryIfNotExists`. If you assume stuff like `AuditSuccess`, `UpdateUptime`, etc.. are already passed into  `RecordAudits`, that removes the need to set those here. And if you no longer need to set those here, you can get rid of the nodes return value.

Reply:
I ended up adding a function called `setAuditFailStatus` which does take an array of nodes instead of just their IDs. The reason for this was because `setAuditFailStatus` is called inside the run function, and it started to get messy to make a new array of just IDs to iterate through.